skimage/feature/peak.py

```diff
@@ -133,60 +175,42 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
             labels[mask] = 1 + rank_order(labels[mask])[0].astype(labels.dtype)
         labels = labels.astype(np.int32)
 
-        # New values for new ordering
-        label_values = np.unique(labels)
-        for label in label_values[label_values != 0]:
-            maskim = (labels == label)
-            out += peak_local_max(image * maskim, min_distance=min_distance,
-                                  threshold_abs=threshold_abs,
-                                  threshold_rel=threshold_rel,
-                                  exclude_border=exclude_border,
-                                  indices=False, num_peaks=num_peaks_per_label,
-                                  footprint=footprint, labels=None)
-
-        # Select highest intensities (num_peaks)
-        coordinates = _get_high_intensity_peaks(image, out, num_peaks)
+        if exclude_border:
+            # create a mask for the non-exclude region
+            inner_mask = _exclude_border(np.ones_like(labels, dtype=bool),
+                                         footprint, exclude_border)
+
+        # For each label, extract a smaller image enclosing the object of
+        # interest, identify num_peaks_per_label peaks and mark them in
+        # variable out.
+        for label_idx, obj in enumerate(ndi.find_objects(labels)):
+            img_object = image[obj] * (labels[obj] == label_idx + 1)
+            mask = _get_peak_mask(img_object, min_distance, footprint,
+                                  threshold_abs, threshold_rel)
+            if exclude_border:
+                # remove peaks fall in the exclude region
+                mask &= inner_mask[obj]
+            coordinates = _get_high_intensity_peaks(img_object, mask,
+                                                    num_peaks_per_label)
+            nd_indices = tuple(coordinates.T)
+            mask.fill(False)
+            mask[nd_indices] = True
+            out[obj] += mask
 
+        coordinates = _get_high_intensity_peaks(image, out, num_peaks)
         if indices is True:
             return coordinates
         else:
             nd_indices = tuple(coordinates.T)
             out[nd_indices] = True
```

Comment:
coordinates is obtained from `out`. Do you need to set `out` again?

Reply:
Good point. If num_peaks is provided, out may contain more than num_peaks candidates.  It needs to be regenerated in that case.

I actually see a bug here.  It was missing a statement             out.fill(False).
The code review really helps!!

It should be:

```
        if not indices and np.isinf(num_peaks):
            return out
        coordinates = _get_high_intensity_peaks(image, out, num_peaks)
        if indices:
            return coordinates
        else:
            out.fill(False)
            nd_indices = tuple(coordinates.T)
            out[nd_indices] = True
            return out

```

Reply:
Where did the first two lines of code come from?

Reply:
Where did the first two lines of code come from?

Reply:
The first two lines is to deal with the case where user does not provide num_peaks (defaults to all, np.inf) and indices is False, out is already the answer that can be returned.
The code I provided (bold) in within the context below:
```

        for i,obj in enumerate(ndi.find_objects(labels)):
            label=i+1
            #print("Label>", label)
            img=image[obj]*(labels[obj]==label)
            mask=_get_peak_mask(img, min_distance, footprint, threshold_abs, threshold_rel)
            if exclude_border:
                # remove peaks fall in the exclude region
                mask &= inner_mask[obj]
            coordinates =_get_high_intensity_peaks(img, mask, num_peaks_per_label)
            nd_indices = tuple(coordinates.T)
            mask.fill(False)
            mask[nd_indices] = True
            out[obj] += mask

        **if not indices and np.isinf(num_peaks):
            return out
        coordinates = _get_high_intensity_peaks(image, out, num_peaks)
        if indices:
            return coordinates
        else:
            out.fill(False)
            nd_indices = tuple(coordinates.T)
            out[nd_indices] = True
            return out**

    # Non maximum filter
    mask = _get_peak_mask(image, min_distance, footprint, threshold_abs, threshold_rel)

```

Reply:
@brownmk are you sure you don't want to open your own PR and take it over from here? You seem to be going through this quite extensively. `Github Desktop`, and editors like `Atom` or `VS Code` make git much less painful.

Reply:
It's less of git, it's the 500-line guideline seems rather intimidating. All issues are addressed now, I will try to learn it next time.

https://github.com/scikit-image/scikit-image/blob/master/CONTRIBUTING.txt

Reply:
That is a really fair point. I'll open an issue about  this.

Most of that is a reference for us to point people to when they hit problems. We often forget how we setup our own environments....

Reply:
@brownmk If you want, I can try guiding you through a setup process in exchange for allowing me to create a PR w/ your code.